src/webgpu/api/validation/copy_between_linear_data_and_texture/copyBetweenLinearDataAndTexture_dataRelated.spec.ts

```diff
@@ -0,0 +1,286 @@
+    const offset = offsetInBlocks * kTextureFormatInfo[format].bytesPerBlock!;
+    const rowsPerImage =
+      copyHeight + rowsPerImagePaddingInBlocks * kTextureFormatInfo[format].blockHeight!;
+    const bytesPerRow =
+      align(t.bytesInACompleteRow(copyWidth, format), bytesPerRowAlignment) +
+      bytesPerRowPadding * bytesPerRowAlignment;
+    const size = { width: copyWidth, height: copyHeight, depth: copyDepth };
+
+    const minDataSize =
+      offset + t.requiredBytesInCopy({ offset, bytesPerRow, rowsPerImage }, format, size);
+
+    // We can't run a failing test with minDataSize = 0.
+    if (minDataSize === 0 && !success) {
+      return;
+    }
+
+    const dataSize = success ? minDataSize : minDataSize - 1;
+
+    const texture = t.createAlignedTexture(format, size);
+
+    t.testRun({ texture }, { offset, bytesPerRow, rowsPerImage }, size, {
+      dataSize,
+      method,
+      success,
+    });
+  });
+
+g.test('texel_block_alignment_on_rows_per_image')
+  .params(
+    params()
+      .combine(poptions('method', kAllTestMethods))
+      .combine(poptions('format', kTextureFormats))
+      .filter(formatCopyableWithMethod)
+      .expand(texelBlockAlignmentTestExpanderForRowsPerImage())
+  )
+  .fn(async t => {
+    const { rowsPerImage, format, method } = t.params;
+    const size = { width: 0, height: 0, depth: 0 };
+
+    const texture = t.createAlignedTexture(format, size);
+
+    const success = rowsPerImage % kTextureFormatInfo[format].blockHeight! === 0;
+
+    t.testRun({ texture }, { bytesPerRow: 0, rowsPerImage }, size, {
+      dataSize: 1,
+      method,
+      success,
+    });
+  });
+
+g.test('texel_block_alignment_on_offset')
+  .params(
+    params()
+      .combine(poptions('method', kAllTestMethods))
+      .combine(poptions('format', kTextureFormats))
+      .filter(formatCopyableWithMethod)
+      .expand(texelBlockAlignmentTestExpanderForOffset())
+  )
+  .fn(async t => {
+    const { format, offset, method } = t.params;
+    const size = { width: 0, height: 0, depth: 0 };
+
+    const texture = t.createAlignedTexture(format, size);
+
+    const success = offset % kTextureFormatInfo[format].bytesPerBlock! === 0;
+
+    t.testRun({ texture }, { offset, bytesPerRow: 0 }, size, { dataSize: offset, method, success });
+  });
+
+g.test('bound_on_bytes_per_row')
+  .params(
+    params()
+      .combine(poptions('method', kAllTestMethods))
+      .combine([
+        { blocksPerRow: 2, additionalBytesPerRow: 0, copyWidthInBlocks: 2 }, // success
+        { blocksPerRow: 2, additionalBytesPerRow: 5, copyWidthInBlocks: 3 }, // success if bytesPerBlock <= 5
+        { blocksPerRow: 1, additionalBytesPerRow: 0, copyWidthInBlocks: 2 }, // failure, bytesPerRow > 0
+        { blocksPerRow: 0, additionalBytesPerRow: 0, copyWidthInBlocks: 1 }, // failure, bytesPerRow = 0
+      ])
+      .combine([
+        { copyHeightInBlocks: 0, copyDepth: 1 }, // we don't have to check the bound
```

Comment:
Should have a test with `copyHeightInBlocks: 1` I think

Reply:
Done